src/NHibernate/Collection/Generic/SetHelpers/SetSnapShot.cs

```diff
@@ -95,13 +123,47 @@ public bool IsReadOnly
 
 		public bool TryGetValue(T element, out T value)
 		{
-			var idx = _elements.IndexOf(element);
-			if (idx >= 0)
+			if (element == null)
+			{
+				// null-s cannot be handled by the cache,
+				// but we keep track of them separately
+				value = default(T);
+				return _containsNull;
+			}
+
+			if (_lookupCache == null)
+			{
+				_lookupCache = new Dictionary<T, T>();
+			}
+
+			if (_lookupCache.TryGetValue(element, out value))
 			{
-				value = _elements[idx];
 				return true;
 			}
 
+			// look at elements not yet in the cache
+			while (_lookupCacheUnprocessed < _elements.Count)
+			{
+				T snapshotElement = _elements[_lookupCacheUnprocessed++];
+				if (snapshotElement != null)
+				{
+					// be careful not to replace cache entries to preserve
+					// original semantics and return the first matching object
+					if (!_lookupCache.ContainsKey(snapshotElement))
+					{
+						_lookupCache.Add(snapshotElement, snapshotElement);
+					}
```

Comment:
This is accidental semantic. Having the snapshot containing two times the same element makes no sense, and would anyway cause bugs if this was actually happening. So it is unneeded.

Reply:
I was not willing to risk that, see above comment.